core/state_transition.go

```diff
@@ -572,3 +651,68 @@ func (st *StateTransition) gasUsed() uint64 {
 func (st *StateTransition) blobGasUsed() uint64 {
 	return uint64(len(st.msg.BlobHashes) * params.BlobTxBlobGasPerBlob)
 }
+
+// distributeTxFees calculates the amounts and recipients of transaction fees and credits the accounts.
+func (st *StateTransition) distributeTxFees() error {
+	if st.evm.Config.NoBaseFee && st.msg.GasFeeCap.Sign() == 0 && st.msg.GasTipCap.Sign() == 0 {
+		// Skip fee payment when NoBaseFee is set and the fee fields
+		// are 0. This avoids a negative effectiveTip being applied to
+		// the coinbase when simulating calls.
+		return nil
+	}
+
+	// Determine the refund and transaction fee to be distributed.
+	refund := new(big.Int).Mul(new(big.Int).SetUint64(st.gasRemaining), st.msg.GasPrice)
+	gasUsed := new(big.Int).SetUint64(st.gasUsed())
+	totalTxFee := new(big.Int).Mul(gasUsed, st.msg.GasPrice)
+	from := st.msg.From
+
+	// Divide the transaction into a base (the minimum transaction fee) and tip (any extra, or min(max tip, feecap - GPM) if espresso).
+	baseTxFee := new(big.Int).Mul(gasUsed, st.calculateGasPriceMinimum())
+	// No need to do effectiveTip calculation, because st.gasPrice == effectiveGasPrice, and effectiveTip = effectiveGasPrice - baseTxFee
+	tipTxFee := new(big.Int).Sub(totalTxFee, baseTxFee)
+
+	feeCurrency := st.msg.FeeCurrency
+	feeHandlerAddress := contracts.FeeHandlerAddress
+
+	log.Trace("distributeTxFees", "from", from, "refund", refund, "feeCurrency", st.msg.FeeCurrency,
+		"coinbaseFeeRecipient", st.evm.Context.Coinbase, "coinbaseFee", tipTxFee,
+		"feeHandler", feeHandlerAddress, "communityFundFee", baseTxFee)
+	if feeCurrency == nil {
+		st.state.AddBalance(st.evm.Context.Coinbase, tipTxFee)
+		st.state.AddBalance(from, refund)
+
+		// Check that we are post bedrock to enable op-geth to be able to create pseudo pre-bedrock blocks (these are pre-bedrock, but don't follow l2 geth rules)
+		// Note optimismConfig will not be nil if rules.IsOptimismBedrock is true
+		if optimismConfig := st.evm.ChainConfig().Optimism; optimismConfig != nil &&
+			st.evm.ChainConfig().IsOptimismBedrock(st.evm.Context.BlockNumber) &&
+			st.evm.ChainConfig().IsCel2(st.evm.Context.Time) {
+			st.state.AddBalance(params.OptimismBaseFeeRecipient, baseTxFee)
+			if cost := st.evm.Context.L1CostFunc(st.evm.Context.BlockNumber.Uint64(), st.evm.Context.Time, st.msg.RollupDataGas, st.msg.IsDepositTx); cost != nil {
+				st.state.AddBalance(params.OptimismL1FeeRecipient, cost)
+			}
+		}
+	} else {
+		// TODO: Send L1 to `OptimismL1FeeRecipient`
+		if err := fee_currencies.CreditFees(st.evm, from, st.evm.Context.Coinbase, feeHandlerAddress, refund, tipTxFee, baseTxFee, feeCurrency); err != nil {
+			log.Error("Error crediting", "from", from, "coinbase", st.evm.Context.Coinbase, "feeHandler", feeHandlerAddress)
+			return err
+		}
+	}
+	return nil
+}
+
+func (st *StateTransition) calculateGasPriceMinimum() *big.Int {
```

Comment:
Maybe worth renaming to avoid GPM connotations from before since we'll be removing it?

Reply:
Renamed it to `calculateBaseFee` and added a comment.